### PR TITLE
mrc-3888: Add prerun endpoint

### PR DESCRIPTION
**Pull request by r-ash**
This builds on #443 

This endpoint will
* Take details of input and output files, check that they exist on disk (will error if they do not). Create an ID for outputs and put into redis. Then format response as project state JSON

### CODE REVIEW

File: R/migrations.R [hunk 1]
```diff
@@ -14,7 +14,7 @@ run_migration <- function(queue, output_dir, dry_run = FALSE) {
   summary_path <- file.path(output_dir, "summary.csv")
   message(sprintf("Saving summary csv %s", summary_path))
   output_path <- file.path(output_dir, "output.qs")
-  write.csv(summary, summary_path, row.names = FALSE)
+  utils::write.csv(summary, summary_path, row.names = FALSE)
```
Comment (r-ash) [inline R/migrations.R:17]: Fixed up some R CMD check notes whilst I was here

File: R/run_model.R [hunk 2]
```diff
@@ -46,7 +46,7 @@ run_model <- function(data, options, path_results, path_prerun = NULL,
     output <- list(
       plot_data_path = NULL,
       model_output_path = system_file("output", "malawi_model_output.qs"),
-      version = packageVersion("naomi"),
+      version = utils::packageVersion("naomi"),
```
Comment (r-ash) [inline R/run_model.R:49]: R CMD note fix

File: R/prerun.R [hunk 3]
```diff
@@ -109,3 +109,91 @@ hash_info_inputs <- function(inputs) {
...
   digest::digest(inputs$md5sum[order(inputs$role)])
 }
+
+prerun <- function(queue) {
+  function(input) {
+    files <- jsonlite::fromJSON(input, simplifyVector = FALSE)
+    all_files <- c(files$inputs, files$outputs)
+    lapply(c(names(all_files)), function(name) {
+      file <- all_files[[name]]
+      if (!file.exists(file$path)) {
```
Comment (richfitz) [inline R/prerun.R:119]: I suspect you'll find it nicer to work with a vectorised version of this that tells you about all missing files at once.

so,


```
msg <- all_files[!file.exists(all_files)]
if (length(msg) > 0) {
  hintr_error(....build a nice error message here...)
}
```

File: R/prerun.R [hunk 4]
```diff
@@ -109,3 +109,91 @@ hash_info_inputs <- function(inputs) {
...
+  model_fit_options <- yaml::read_yaml(text = output$info$options.yml)
+  calibration_options <- yaml::read_yaml(text =
+                                           output$info$calibration_options.yml)
+  packages <- utils::read.csv(text = output$info$packages.csv)
+
+  inputs <- build_state_inputs(inputs)
+  fit <- build_state_output(queue, model_fit_output, model_fit_options)
+  calibrate <- build_state_output(queue, calibrate_output, calibration_options)
+  version <- build_state_version(packages[packages$name == "naomi", "version"])
+  state <- list(
```
Comment (richfitz) [inline R/prerun.R:153]: stray assignment here, left over from refactoring?

File: R/prerun.R [hunk 5]
```diff
@@ -109,3 +109,91 @@ hash_info_inputs <- function(inputs) {
...
+
+prerun_build_state <- function(queue, inputs, model_fit_output,
+                               calibrate_output) {
+  output <- naomi::read_hintr_output(calibrate_output$model_output_path)
+  model_fit_options <- yaml::read_yaml(text = output$info$options.yml)
+  calibration_options <- yaml::read_yaml(text =
+                                           output$info$calibration_options.yml)
+  packages <- utils::read.csv(text = output$info$packages.csv)
+
+  inputs <- build_state_inputs(inputs)
```
Comment (richfitz) [inline R/prerun.R:149]: not really following the "build" naming here?

File: R/prerun.R [hunk 6]
```diff
@@ -109,3 +109,91 @@ hash_info_inputs <- function(inputs) {
...
+  fit <- build_state_output(queue, model_fit_output, model_fit_options)
+  calibrate <- build_state_output(queue, calibrate_output, calibration_options)
+  version <- build_state_version(packages[packages$name == "naomi", "version"])
```
Comment (richfitz) [inline R/prerun.R:152]:
```
packages$version[packages$name == "naomi"]
```

relies less on terrible data frame behaviour

File: R/prerun.R [hunk 7]
```diff
@@ -109,3 +109,91 @@ hash_info_inputs <- function(inputs) {
...
+    version = version
+  )
+}
+
+build_state_inputs <- function(inputs) {
+  lapply(inputs, function(input) {
+    ## web app expects path like as uploads/file_name.csv
+    ## if leading / is included then it takes "uploads" as the filename and
+    ## errors
+    if (substring(input$path, 1, 1) == "/") {
```
Comment (richfitz) [inline R/prerun.R:166]:
```
path <- sub("^/", "", input$path)
```

can be done unconditionally

File: R/prerun.R [hunk 8]
```diff
@@ -109,3 +109,91 @@ hash_info_inputs <- function(inputs) {
...
+  id <- create_result(queue, output)
+  list(
+    options = recursive_scalar(options),
+    id = scalar(id)
+  )
+}
+
+create_result <- function(queue, result) {
+  task_id <- ids::random_id()
+  rrq:::run_task_cleanup(queue$queue$con,
```
Comment (richfitz) [inline R/prerun.R:186]: hah, is that all? That's pretty cute

File: R/prerun.R [hunk 9]
```diff
@@ -109,3 +109,91 @@ hash_info_inputs <- function(inputs) {
...
+                         queue$queue$.__enclos_env__$private$store,
+                         task_id,
+                         rrq:::TASK_COMPLETE,
+                         result)
+  task_id
+}
+
+build_state_version <- function(naomi_version) {
+  version <- cfg$version_info
+  version$naomi <- scalar(as.character(naomi_version))
```
Comment (richfitz) [inline R/prerun.R:197]: with the change above you might be able to drop this `as.character` here, unless I am missing something....


Where is `cfg` coming from? is that global?

**Review by richfitz — CHANGES_REQUESTED** [review]
OK, small changes suggested - but at some point we *really* need some docs on this

**Comment by r-ash** [comment]
> OK, small changes suggested - but at some point we _really_ need some docs on this

So I have a write up in the naomi wiki on how prerun works atm https://github.com/mrc-ide/hintr/pull/444#pullrequestreview-1219212328 can update this there but could also put in a vignette or something

File: R/prerun.R (hunk 5 above)
Comment (r-ash) [reply R/prerun.R:149]: Constructing the state json, from inputs, model fit, calibrate, version etc.

File: R/prerun.R (hunk 9 above)
Comment (r-ash) [reply R/prerun.R:197]: Yeah cfg is global, defined in `zzz.R`

File: R/prerun.R (hunk 8 above)
Comment (r-ash) [reply R/prerun.R:187]: Yes not too bad in the end at all!

File: R/prerun.R (hunk 3 above)
Comment (r-ash) [reply R/prerun.R]: Tidied this up now

**Review by richfitz — APPROVED** [review]

**Pull request merged**